include/revng/Pipeline/ContainerInspector.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/ManagedStatic.h"
+
+#include "revng/Pipeline/Containers.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Target.h"
+
+namespace Pipeline {
+
+/// If a downstream user of reveng wishes to extend the enumeration and
+/// removal rule from a container it can do so by extending this class,
+/// specialized with the type of the container in question.
+///
+/// the enumerate and removeTarget methods must be implemented with the
+/// appropriate rules
+///
+///
+template<typename Container>
+class ContainerInspector {
+private:
+  Kind *K;
+
+public:
+  using StaticContainer = llvm::SmallVector<ContainerInspector *, 4>;
+
+  ContainerInspector(Kind &K) : K(&K) {
+    getRegisteredInspectors().push_back(this);
+  }
+  virtual ~ContainerInspector() = default;
+
+  virtual TargetsList
+  enumerate(const Context &Ctx, const Container &ToInspect) const = 0;
+  virtual bool remove(const Context &Ctx,
+                      const TargetsList &Targets,
+                      Container &ToInspect) const = 0;
+
+  static TargetsList
+  enumerateTargets(const Context &Ctx, const Container &ToInspect) {
+    TargetsList ToReturn;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      ToReturn.merge(Inspector->enumerate(Ctx, ToInspect));
+
+    return ToReturn;
+  }
+
+  static bool removeTarget(const Context &Ctx,
+                           const TargetsList &Targets,
+                           Container &ToInspect) {
+    bool RemovedAll = true;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      RemovedAll = Inspector->remove(Ctx,
+                                     Targets.filter(*Inspector->K),
+                                     ToInspect)
+                   and RemovedAll;
+
+    return RemovedAll;
+  }
+
+private:
+  static StaticContainer &getRegisteredInspectors() {
+    static StaticContainer List;
+    return List;
+  }
```

Comment:
Is this to avoid having to declare the instantiation of the `static` field?

Reply:
it's avoiding me to have to provide a cpp file with only that static member inside